keps/sig-cli/kustomize-annotate-generated.md

```diff
@@ -0,0 +1,119 @@
+- Secret and ConfigMap annotations are retained after the Workloads stop referencing them
+
+## Motivation
+
+- Kustomize generates Secret and ConfigMap Resources which are referenced by Workloads
+  - When the data for the Secrets or ConfigMaps change, new Secret or ConfigMap Resources are generated
+  - Kustomize updates the Workloads to refer to the new Resources, and stop referring to the old Resources
+  - References to the old Secret and ConfigMap Resources are leaked when new Resources are generated
+  - The information that the Resources were generated by Kustomize rather than explicitly defined by
+    users is lost
+
+### Goals
+
+- Provide provenance for Resources generated by Kustomize
+  - Retain a record of Secret and ConfigMap Resources generated by Kustomize.
+  - Retain a record of which Workloads each generated Secret and ConfigMap was consumed by.
+  - Retain a record of which generated Secret and ConfigMap Resources each Workload consumed.
+- Allow other tools to identify generated Secrets and ConfigMaps that are no longer in use as
+  candidates for garbage collection.
+
+### Non-Goals
+
+## Proposal
+
+Kustomize already modifies Workloads that refer to generated Secrets and ConfigMaps by updating
+the names to match the generated names.
+
+### Annotations on Workloads
+
+Set an annotation on each Workload for each generated Secret and ConfigMap it references.  When
+using Apply, these annotations will be retained even after Kustomize stops generating them.
+
+**Secret Annotation:** `generated-secret.kustomize.sigs.k8s.io/<generated-secret--name>=`
+**ConfigMap Annotation:** `generated-configmap.kustomize.sigs.k8s.io/<generated-configmap-name>=`
+
+### Annotations on Generated Secret and ConfigMap Resources  
+
+Set an annotation on each generated Secret and ConfigMap Resource for each Workload that
+consumes it.  These annotations will be retained even after Kustomize stops generating them.
+
+**Secret Annotation:** `secret-reference.kustomize.sigs.k8s.io/<workload-group>.<workload-kind>.<workload-name>=`
+**ConfigMap Annotation:** `configmap-reference.kustomize.sigs.k8s.io/<workload-group>.<workload-kind>.<workload-name>=`
+
+Generated Secret and ConfigMap Resources should also have an annotation stating that
+they were generated.
+
+**Annotation:** `kustomize.sigs.k8s.io/generated=`
+
+
+### Risks and Mitigations
+
+Because the list of annotations don't get deleted, they will grow over time.  This is also
+true of the generated Resources themselves.  This can be mitigated by having the process
+that cleans up Secrets and ConfigMaps remove the annotations from the Workloads.
+
+This behavior can be disabled by setting `annotateProvenance: false`.
+
+## Graduation Criteria
+
+NA
+
+## Docs
+
+Update Kustomize generateSecret and generateConfigMap documentation
+
+## Test plan
+
+### Version Skew Tests
+
+NA
+
+## Implementation History
+
+## Alternatives
```

Comment:
An alternative would be to solve this on the server-side, perhaps by either:
- adding a volume type that allowed inline content (ie: the configmap content would be embedded directly in the PodSpec) - not appropriate for Secrets.
- adding a generic reference counting mechanism, so objects could declare that they should not be deleted while another object/selector exists.  I haven't thought hard enough about this yet to work out whether existing garbage collection, finalizers, or owner refs functionality can be turned into this.

Point is: this is not a problem unique to kustomize.  As a part of the k8s project, we should at least _think_ about a way to solve this for everyone...

Reply:
There have been proposals to do so in the past, but none have been accepted.  I will try to dig them up...

Reply:
@janetkuo proposed a server-side solution, but it was closed: kubernetes/community#2287

Reply:
So: could we include a pointer to that KEP, with a brief one-line-summary of why that approach was not acceptable?  (after a quick look, I can't work out where that KEP was moved _to_.  Was it actually rejected, or just lost in a paper shuffle?)

Reply:
Yes absolutely.  I am a bit backlogged right now, but will update the KEP with more context.  It's not clear to me this is the right way to go, but I thought I'd throw it out for discussion.